hibernate-core/src/main/java/org/hibernate/query/sqm/tuple/internal/AnonymousTupleType.java

```diff
@@ -110,17 +112,33 @@ public String getTypeName() {
 		return SqmDomainType.super.getTypeName();
 	}
 
-	private static SqmSelectableNode<?>[] extractSqmExpressibles(SqmSubQuery<?> subQuery) {
-		final SqmSelectClause selectClause = subQuery.getQuerySpec().getSelectClause();
+	private static SqmSelectableNode<?>[] extractSqmExpressibles(SqmSelectQuery<?> selectQuery) {
+		final SqmSelectClause selectClause = selectQuery.getQueryPart()
+				.getFirstQuerySpec()
+				.getSelectClause();
 		if ( selectClause == null || selectClause.getSelectionItems().isEmpty() ) {
-			throw new IllegalArgumentException( "subquery has no selection items" );
+			throw new IllegalArgumentException( "selectQuery has no selection items" );
 		}
-		// todo: right now, we "snapshot" the state of the subquery when creating this type, but maybe we shouldn't?
-		//  i.e. what if the subquery changes later on? Or should we somehow mark the subquery to signal,
+		// todo: right now, we "snapshot" the state of the selectQuery when creating this type, but maybe we shouldn't?
+		//  i.e. what if the selectQuery changes later on? Or should we somehow mark the selectQuery to signal,
 		//  that changes to the select clause are invalid after a certain point?
 		return selectClause.getSelectionItems().toArray( SqmSelectableNode[]::new );
 	}
 
+	private static String[] extractAliases(SqmSelectQuery<?> selectQuery) {
+		final SqmSelectClause selectClause = selectQuery.getQueryPart()
+				.getFirstQuerySpec()
+				.getSelectClause();
+		final List<String> aliases = new ArrayList<>();
+		for (final SqmSelection<?> selection : selectClause.getSelections()) {
+			final String alias = selection.getAlias();
+			selection.getSelectableNode().visitSubSelectableNodes( node ->
+					aliases.add( alias == null ? node.getAlias() : alias )
```

Comment:
Can you explain why you needed to do this? It's still a bit obscure to me why you'd fall back to the parent node's alias, especially if there's more than one sub-selectable node that's seems outright not correct.

Reply:
It may looks ugly and confusing, but it actually, works. `selection.getAlias()` will be null if (and only if) `selection.getSelectableNode()` is compound selection. This can be replaced with cleaner:

```
			if ( selection.getSelectableNode().isCompoundSelection() ) {
				selection.getSelectableNode().visitSubSelectableNodes( node ->
						aliases.add( node.getAlias() )
				);
			}
			else {
				aliases.add( selection.getAlias() );
			}
```